### PR TITLE
docs: Expands docs around C function pointers

### PR DESCRIPTION
**Pull request by TimDeve**
The PR enables html docs generation for the `Function` module as well as expand the docs around its usage.

### CODE REVIEW

File: core/Function.carp [hunk 1]
```diff
@@ -1,7 +1,16 @@
 (defmodule Function
-  (doc unsafe-ptr "returns void pointer to the function passed in."
-                 "This is unsafe as unsafe-ptr can't check the value passed in is a function.")
+  (doc unsafe-ptr "returns a void pointer to the function passed in.
+
+  This is unsafe as unsafe-ptr can't check the value passed in is a function.
+
+  It is the responsability of the caller to ensure that the lifetime of the pointer
```
Comment (scolsen) [inline core/Function.carp:6]: typo s/responsability/responsibility

File: core/Function.carp [hunk 2]
```diff
@@ -1,7 +1,16 @@
...
+  does not exceed the lifetime of the function it's used on.")
   (deftemplate unsafe-ptr (Fn [(Ref a)] (Ptr ())) "void* $NAME($a *fn)" "$DECL { return fn->callback; }")
-  (doc unsafe-env-ptr "returns void pointer to the environment captured by a lambda."
-                   "This is unsafe as unsafe-env-ptr can't check the value passed in is a function.")
+
+  (doc unsafe-env-ptr "returns  void pointer to the environment captured by a lambda.
+
+  This is unsafe as unsafe-env-ptr can't check the value passed in is a function.
+
+  It is the responsability of the caller to ensure that the lifetime of the pointer
```
Comment (scolsen) [inline core/Function.carp:14]: typo s/responsability/responsibility

File: docs/CInterop.md [hunk 3]
```diff
@@ -342,3 +343,44 @@ out/main.c:9153:29: error: invalid operands to binary expression ('String' (aka
...
+## Callbacks
+
+Some C APIs rely on callbacks, let's define a C function that accepts a
+callback and an argument and returns the result of calling that function as an
+example:
+
+```clojure
+(deftemplate runner (Fn [(Ptr ()) (Ptr ())] a)
+                    "$a $NAME(void* fn, void* args)"
+                    "$a $NAME(void* fnptr, void* args) {
```
Comment (scolsen) [inline docs/CInterop.md:355]: Maybe make the argument names for the function pointer the same in the signature string and declaration?

File: docs/CInterop.md [hunk 4]
```diff
@@ -342,3 +343,44 @@ out/main.c:9153:29: error: invalid operands to binary expression ('String' (aka
...
+environment.
+
+In the second example we want to use a static function so we can use
+`Function.unsafe-ptr` again, and the argument we pass in needs to be coerced
+from a `Ref` into a `(Ptr ())`.
+
+Because everything gets turned into a void pointer all type safety is lost so
+it is the responsibility of the caller to ensure the operation is safe. It is
+also important to ensure the lifetime of the `Ptr` doesn't not exceed the
+lifetime of the function/env it represent.
```
Comment (scolsen) [inline docs/CInterop.md:385]: s/represent/represents

**Review by scolsen — APPROVED** [review]
Wonderful!

File: core/Function.carp (hunk 1 above)
Comment (TimDeve) [reply core/Function.carp:6]: French is leaking through 🙂

File: docs/CInterop.md (hunk 3 above)
Comment (TimDeve) [reply docs/CInterop.md:355]: Not sure why I didn't use `$DECL` here really.

**Comment by TimDeve** [comment]
Fixed the typos.

**Pull request merged**